src/Lecture4.hs

```diff
@@ -237,7 +294,19 @@ you can return string "no value"
 -}
 
 displayStats :: Stats -> String
-displayStats = error "TODO"
+displayStats stats = unlines [
+  "Total positions:       : " ++ show (getSum $ statsTotalPositions stats),
+  "Total final balance    : " ++ show (getSum $ statsTotalSum stats),
+  "Biggest absolute cost  : " ++ show (getMax $ statsAbsoluteMax stats),
+  "Smallest absolute cost : " ++ show (getMin $ statsAbsoluteMin stats),
+  "Max earning            : " ++ showMaybe getMax (statsSellMax stats),
+  "Min earning            : " ++ showMaybe getMin (statsSellMin stats),
+  "Max spending           : " ++ showMaybe getMax (statsBuyMax stats),
+  "Min spending           : " ++ showMaybe getMin (statsBuyMin stats),
+  "Longest product name   : " ++ unMaxLen (statsLongest stats)
+  ]
+  where
+    showMaybe = \fn m -> if isNothing m then "no value" else show $ fn $ fromJust m
```

Comment:
The combination of `isNothing` and `fromJust` is considered an anti-pattern.
Here you can use standard function `maybe` instead to write this in a more elegant way:

```suggestion
    showMaybe fn = maybe "no value" (show . fn)
```